client/state/utils/with-enhancers.ts

```diff
@@ -9,22 +19,25 @@
  * @see client/state/analytics/actions/enhanceWithSiteType for an example
  * @see extendAction from @automattic/state-utils for a simpler alternative
  */
-export const withEnhancers = ( actionCreator, enhancers ) => ( ...args ) => (
-	dispatch,
-	getState
-) => {
+export const withEnhancers = < TActionArguments extends any[] >(
+	actionCreator: ( ...args: TActionArguments ) => Action | AnyThunkAction,
+	enhancers: Enhancer | Enhancer[]
```

Comment:
What if we used this opportunity to make the somewhat loose API more strict? TypeScript is nudging us here to simplify the type of `enhancers` and make the `withEnhancers` signature either
```js
withEnhances( actionCreator, enhancers: Enhancers[] );
```
or
```js
withEnhancers( actionCreator, ...enhancers: Enhancers[] );
```

Reply:
Also, let's look at #51889 and consider merging it first. I suspect it might simplify some types inside the `withEnhancers` implementation.